### DISCUSSION TITLE
Some questions about vs project generator

### DISCUSSION THREAD

**Discussion by xq114** [xmake-io/xmake#2087] [General]
https://github.com/xmake-io/xmake/blob/master/xmake/plugins/project/vstudio/vs.lua#L31

这里为啥使用config.get而不用core.tool.toolchain.load("vs")？

https://github.com/xmake-io/xmake/blame/master/xmake/plugins/project/vstudio/impl/vs201x.lua#L528

这里为何要清空缓存？如果想要生成项目时使用特定的option，应该怎么处理？

**Comment by waruqi** [comment]
> https://github.com/xmake-io/xmake/blob/master/xmake/plugins/project/vstudio/vs.lua#L31
> 
> 这里为啥使用config.get而不用core.tool.toolchain.load("vs")？

搞这个时候，还没 toolchain 模块了，不过这里用的就是全局 vs 配置，直接用 config.get 也足够了，当然，用 toolchain 的也可以。

> 
> https://github.com/xmake-io/xmake/blame/master/xmake/plugins/project/vstudio/impl/vs201x.lua#L528
> 
> 这里为何要清空缓存？如果想要生成项目时使用特定的option，应该怎么处理？

所有 cache 都得清，否则切 arch/mode 会收到各种cache干扰

生成项目的 option ，在执行 xmake config 阶段，应该都 merge 进 configs 里面去了，这里 option cache 不会用到。 config.get 里面没去清

**Reply by xq114** [reply]
https://github.com/xmake-io/xmake/blob/master/xmake/plugins/project/vstudio/impl/vs201x.lua#L520-L523

这一段的作用是什么

**Reply by waruqi** [reply]
清掉 option() 定义选项的 状态值，强制走一遍 option check ，比如 func, include, types 检测什么的

**Comment by xq114** [comment]
target:sourcekinds()函数是怎么一回事？为什么我在cmake generator中使用这个函数能正常运行，在vs generator中用就报错
```
error: @programdir\core\main.lua:280: @programdir\plugins\project\vstudio\impl\vs201x_vcxproj.lua:46: method 'sourcekinds' is not callable (a nil value)
stack traceback:
    [@programdir\plugins\project\vstudio\impl\vs201x_vcxproj.lua:46]: in function '_check_cuda'
    [@programdir\plugins\project\vstudio\impl\vs201x_vcxproj.lua:243]: in function '_make_configurations'
    [@programdir\plugins\project\vstudio\impl\vs201x_vcxproj.lua:1058]: in function 'make'
    [@programdir\plugins\project\vstudio\impl\vs201x.lua:588]: in function 'make'
    [@programdir\plugins\project\vstudio\vs.lua:50]: in function '?'
    [@programdir\plugins\project\main.lua:79]: in function '_make'
    [@programdir\plugins\project\main.lua:89]:
    [C]: in function 'xpcall'
    [@programdir\core\base\utils.lua:280]:
    [@programdir\core\base\task.lua:519]: in function 'run'
    [@programdir\core\main.lua:278]: in function 'cotask'
    [@programdir\core\base\scheduler.lua:371]:

stack traceback:
        [C]: in function 'error'
        @programdir\core\base\os.lua:849: in function 'os.raiselevel'
        (...tail calls...)
        @programdir\core\main.lua:280: in upvalue 'cotask'
        @programdir\core\base\scheduler.lua:371: in function <@programdir\core\base\scheduler.lua:368>
```

**Reply by waruqi** [reply]
这里的 target 不是 project target 对象。只是名字相同而已 实际是这个

https://github.com/xmake-io/xmake/blob/d1628f8c57ecc60823a5d0b59bf070cddc794443/xmake/plugins/project/vstudio/impl/vs201x.lua#L557

**Reply by xq114** [reply]
为啥这里_target.info要做成一个table？里面明明只有一个元素

**Reply by waruqi** [reply]
target info 不是好多东西了，哪来只有一个元素

https://github.com/xmake-io/xmake/blob/d1628f8c57ecc60823a5d0b59bf070cddc794443/xmake/plugins/project/vstudio/impl/vs201x.lua#L236

**Comment by xq114** [comment]
https://github.com/xmake-io/xmake/blob/master/xmake/plugins/project/cmake/cmakelists.lua#L81-L95
https://github.com/xmake-io/xmake/blob/master/xmake/plugins/project/vstudio/impl/vs201x.lua#L291
https://github.com/xmake-io/xmake/blob/master/xmake/plugins/project/vsxmake/getinfo.lua#L123-L129

为啥几个generator获取compile flag的方式全都不一样？要获取cuda编译的flag应该以哪个为准？compiler.compflags这个接口还适用吗？

**Reply by waruqi** [reply]
> 为啥几个generator获取compile flag的方式全都不一样？要获取cuda编译的flag应该以哪个为准？compiler.compflags这个接口还适用吗？

这三个地方用途不同，看具体需求而定

cmake, getinfo 这两不需要取完整所有 flags，只是针对性提取里面的 add_cflags，add_defines add_links 的配置对应的 flags而已

而 vs201x 那个需要取整个 编译命令行的 flags ，并且不需要根据根据指定 links defines 作为 key ，挨个提取，用法不同，都有各自用途
